orttraining/orttraining/core/optimizer/scaled_sum_fusion.cc

```diff
@@ -0,0 +1,256 @@
+      continue;
+    }
+
+    InlinedVector<NodeArg*> data_input_args;
+    InlinedVector<float> scales;
+    data_input_args.reserve(3);
+    scales.reserve(3);
+    InlinedVector<std::reference_wrapper<Node>> nodes_to_remove;
+
+    // Be noted: it is possible the two input nodes are from the same node.
+    const Node* add_input_0 = graph_utils::GetInputNode(node, 0);
+    const Node* add_input_1 = graph_utils::GetInputNode(node, 1);
+    if (add_input_0 == nullptr || add_input_1 == nullptr) {
+      continue;
+    }
+
+    auto check_add_input = [&graph, &output_shape, &nodes_to_remove,
+                            &data_input_args, &scales](Node* add_input_node) -> bool {
+      float scale_value = 1.0f;
+      if (!IsScaleOperator(graph, *add_input_node, output_shape, scale_value)) {
+        return false;
+      }
+
+      auto it = std::find_if(nodes_to_remove.begin(), nodes_to_remove.end(),
+                             [&add_input_node](std::reference_wrapper<Node> n) {
+                               return ((Node&)n).Index() == add_input_node->Index();
+                             });
+      if (it == nodes_to_remove.end()) {
+        nodes_to_remove.push_back(*add_input_node);
+      }
+
+      data_input_args.push_back(add_input_node->MutableInputDefs()[0]);
+      scales.push_back(scale_value);
+
+      return true;
+    };
+
+    Node* add_input_node_0 = graph.GetNode(add_input_0->Index());
+    Node* add_input_node_1 = graph.GetNode(add_input_1->Index());
+    if (!check_add_input(add_input_node_0) || !check_add_input(add_input_node_1)) {
+      continue;
+    }
+
+    Node* last_node = &node;
+    // Handle three inputs.
+    if (node.GetOutputEdgesCount() == 1) {
+      Node& output_node = *graph.GetNode(node.OutputEdgesBegin()->GetNode().Index());
+      int output_node_port = node.OutputEdgesBegin()->GetDstArgIndex();
+      if (graph_utils::IsSupportedOptypeVersionAndDomain(output_node, "Add", {6, 7, 13, 14}) &&
+          !graph.IsOutput(output_node.OutputDefs()[0]) /*this Add cannot generate graph output */
+      ) {
+        int the_other_input_port = 1 - output_node_port;
+        NodeArg* the_other_input_arg = output_node.MutableInputDefs()[the_other_input_port];
+        const Node* the_other_input_node = graph.GetProducerNode(the_other_input_arg->Name());
+        Node* mutable_the_other_input_node = the_other_input_node
+                                                 ? graph.GetNode(the_other_input_node->Index())
+                                                 : nullptr;
+
+        bool the_other_node_output_edge_check = mutable_the_other_input_node == nullptr ||
+                                                mutable_the_other_input_node->GetOutputEdgesCount() == 1;
+
+        if (the_other_node_output_edge_check &&
+            the_other_input_arg->Shape() && IsShapeEqual(the_other_input_arg->Shape(), output_shape)) {
+          last_node = &output_node;
+          nodes_to_remove.push_back(node);
```

Comment:
In case of mutable_the_other_input_node is Initializer or not Div node, do we also need to push arg into data_input_args?